File: Programacion I/24_ejercicio_parcial_2/mutant_detector.py
class MutantDetector:
    def __init__(self):
        # Inicializa una lista vacía para almacenar la matriz de ADN.
        self.dna_matrix = []
        # Contador para secuencias de cuatro letras iguales encontradas.
        self.sequences_found = 0

    def get_dna_from_user(self):
        # Solicita al usuario que ingrese las secuencias de ADN.
        print(
            "\nIngrese las secuencias de ADN (6 secuencias de 6 letras cada una de A, T, C, G):"
        )
        # Reinicia la matriz de ADN para recibir nuevas entradas.
        self.dna_matrix = []
        # Itera seis veces, una por cada secuencia de ADN necesaria.
        for i in range(6):
            # Un bucle que sigue solicitando la secuencia hasta que sea válida.
            while True:
                # Pide al usuario una secuencia y la convierte a mayúsculas.
                sequence = input(f"Secuencia {i+1}: ").upper()
                # Verifica que la secuencia tenga 6 caracteres y sean todos válidos (A, T, C, G).
                if len(sequence) == 6 and all(char in "ATCG" for char in sequence):
                    # Agrega la secuencia válida a la matriz de ADN.
                    self.dna_matrix.append(sequence)
                    # Rompe el bucle para pasar a la siguiente secuencia.
                    break
                # Imprime un mensaje de error si la secuencia es inválida.
                print(
                    "Secuencia inválida. Asegúrese de que sean 6 caracteres de A, T, C, G."
                )
        # Confirma que la matriz se ha cargado.
        print("\nMatriz de ADN cargada correctamente.")

    def is_mutant(self, dna):
        # Verifica si la matriz de ADN es válida.
        if not self._is_valid_dna_matrix(dna):
            # Retorna False si la matriz no es válida.
            return False
        # Almacena la matriz de ADN proporcionada en la variable de instancia.
        self.dna_matrix = dna
        # Reinicia el contador de secuencias encontradas.
        self.sequences_found = 0
        # Obtiene el tamaño de la matriz.
        n = len(dna)

        # Recorre cada elemento de la matriz.
        for i in range(n):
            for j in range(n):
                # Verifica si ya se encontraron más de una secuencia.
                if self.sequences_found > 1:
                    # Retorna True si es así, ya que se confirma que es mutante.
                    return True

                # Inicia la búsqueda en profundidad desde la celda actual.
                if self._dfs(i, j, -1, self.dna_matrix[i][j], 0):
                    # Incrementa el contador si se encuentra una secuencia.
                    self.sequences_found += 1

        # Retorna True si se encontraron más de una secuencia, False en caso contrario.
        return self.sequences_found > 1

    def _is_valid_dna_matrix(self, dna):
        # Verifica que la matriz tenga 6 filas y cada fila tenga 6 caracteres.
        if len(dna) != 6 or any(len(row) != 6 for row in dna):
            return False
        # Verifica que todos los caracteres sean válidos (A, T, C, G).
        if any(char not in "ATCG" for row in dna for char in row):
            return False
        # Retorna True si todas las validaciones son correctas.
        return True

    def _dfs(self, i, j, direction, current_char, depth):
        # Caso base para cuando estamos comenzando la búsqueda en todas las direcciones.
        if direction == -1 and depth == 0:
            # Llama a _dfs en todas las direcciones desde la celda actual.
            return any(self._dfs(i, j, d, current_char, 1) for d in range(8))

        # Verifica los límites y si el carácter actual coincide con el buscado.
        if (
            i < 0
            or i >= len(self.dna_matrix)
            or j < 0
            or j >= len(self.dna_matrix)
            or self.dna_matrix[i][j] != current_char
        ):
            return False

        # Si hemos alcanzado la profundidad de 4, hemos encontrado una secuencia completa.
        if depth == 4:
            return True

        # Define las 8 posibles direcciones de movimiento (i, j).
        directions = [
            (-1, 0),  # Arriba
            (1, 0),  # Abajo
            (0, -1),  # Izquierda
            (0, 1),  # Derecha
            (-1, -1),  # Diagonal arriba izquierda
            (-1, 1),  # Diagonal arriba derecha
            (1, -1),  # Diagonal abajo izquierda
            (1, 1),  # Diagonal abajo derecha
        ]

        # Asegura que la dirección es válida antes de hacer la llamada recursiva.
        if 0 <= direction < len(directions):
            next_i, next_j = i + directions[direction][0], j + directions[direction][1]
            # Realiza la llamada recursiva para el siguiente nodo en la dirección dada.
            return self._dfs(next_i, next_j, direction, current_char, depth + 1)

        # Retorna False si la dirección no es válida (fuera de los límites de la matriz).
        return False

File: Primer Semestre/Laboratorio de Computación I/MutantDetector.java
import java.util.HashSet;
import java.util.Scanner;
import java.util.Set;
/* cSpell:disable */

/*
 * EXAMEN PARCIAL
 * Magneto quiere reclutar la mayor cantidad de mutantes para poder luchar
 * contra
 * los X-Mens.
 * Te ha contratado a ti para que desarrolles un programa que detecte si un
 * humano
 * es mutante basándose en su secuencia de ADN.
 * 
 * Para eso te ha pedido crear un programa con un método o función booleana,
 * llamada is_mutant(dna), que recibe como parámetro un arreglo de Strings que
 * representan cada fila de una matriz 6x6 con la secuencia de ADN.
 * Las letras de los Strings solo pueden ser: A,T,C,G; las cuales representan
 * cada
 * base nitrogenada del ADN.
 * 
 * Sabrás si un humano es mutante, si encuentras más de una secuencia de cuatro
 * letras iguales, estas pueden aparecer de forma oblicua, horizontal o
 * vertical.
 * Ejemplo de matrices válidas:
 * 
 * No-Mutante:
 * A T G C G A
 * C A G T G C
 * T T A T T T
 * A G A C G G
 * G C G T C A
 * T C A C T G
 * Mutante:
 * A T G C G A
 * C A G T G C
 * T T A T G T
 * A G A A G G
 * C C C C T A
 * T C A C T G
 * 
 * En el caso de la segunda matriz ,
 * adn = [‘ATGCGA’,’CAGTGC’,’TTATGT’,’AGAAGG’,’CCCCTA’,’TCACTG’]
 * la función devolverá True.
 * 
 * Desarrolla el algoritmo de forma más eficiente posible.
 * El ingreso de los valores de la matriz debe ser pedido por teclado. Ten en
 * cuenta casos para que sea mutante y casos en los que no lo sea.
 * Una vez cargada correctamente la misma (esto debe verificarse), aplique la
 * función que verifica si hay presencia de genes mutantes o no y mostrar el
 * resultado por pantalla al usuario.
 * 
 * Subir a Github el proyecto con los casos de prueba.
 */

class MutantDetector {

    // Matriz para almacenar el ADN
    private char[][] dnaMatrix;

    // Contador de secuencias encontradas
    private int sequencesFound;

    // Conjunto para almacenar letras únicas de secuencias encontradas
    private Set<Character> uniqueLettersFound;

    // Constructor
    public MutantDetector() {
        dnaMatrix = new char[6][6];
        sequencesFound = 0;
        uniqueLettersFound = new HashSet<>();
    }

    // Método para obtener el ADN del usuario
    public void getDnaFromUser() {

        Scanner sc = new Scanner(System.in);

        System.out.println("Ingrese las secuencias de ADN (6 secuencias de 6 letras cada una de A, T, C, G):");
        for (int i = 0; i < 6; i++) {
            while (true) {
                System.out.print("Secuencia " + (i + 1) + ": ");
                String sequence = sc.nextLine().toUpperCase();
                if (isValidSequence(sequence)) {
                    dnaMatrix[i] = sequence.toCharArray();
                    break;
                }
                System.out.println("Secuencia inválida. Asegúrese de que sean 6 caracteres de A, T, C, G.");
            }
        }

        sc.close();

        System.out.println("\nMatriz de ADN cargada correctamente.");
    }

    // Método principal para determinar si un humano es mutante
    public boolean isMutant(char[][] dna) {
        if (!isValidDnaMatrix(dna)) {
            return false;
        }

        dnaMatrix = dna;
        sequencesFound = 0;
        uniqueLettersFound.clear();

        int n = dna.length;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (sequencesFound > 1) {
                    return true;
                }

                if (dfs(i, j, -1, dnaMatrix[i][j], 0)
                        && dnaMatrix[i][j] != uniqueLettersFound.iterator().next()) {
                    sequencesFound++;
                }
            }
        }

        return sequencesFound > 1;
    }

    // Método para validar si una matriz de ADN es válida
    private boolean isValidDnaMatrix(char[][] dna) {
        if (dna.length != 6 || anyRowLengthNotEqualSix(dna) || anyInvalidCharacter(dna)) {
            return false;
        }
        return true;
    }

    // Método para verificar si alguna fila no tiene longitud igual a seis
    private boolean anyRowLengthNotEqualSix(char[][] dna) {
        for (char[] row : dna) {
            if (row.length != 6) {
                return true;
            }
        }
        return false;
    }

    // Método para verificar si algún carácter es inválido
    private boolean anyInvalidCharacter(char[][] dna) {
        for (char[] row : dna) {
            for (char c : row) {
                if (c != 'A' && c != 'T' && c != 'C' && c != 'G') {
                    return true;
                }
            }
        }
        return false;
    }

    // Método para validar si una secuencia es válida
    private boolean isValidSequence(String sequence) {
        return sequence.length() == 6 && sequence.matches("[ATCG]+");
    }

    // Método de búsqueda en profundidad (DFS)
    private boolean dfs(int i, int j, int direction, char currentChar, int depth) {
        if (direction == -1 && depth == 0) {
            // Llamada a DFS en todas las direcciones posibles
            for (int d = 0; d < 8; d++) {
                if (dfs(i, j, d, currentChar, 1)) {
                    return true;
                }
            }
            return false;
        }

        if (i < 0 || i >= dnaMatrix.length || j < 0 || j >= dnaMatrix.length || dnaMatrix[i][j] != currentChar) {
            return false;
        }

        if (depth == 4) {
            char currentLetter = dnaMatrix[i][j];
            if (!uniqueLettersFound.contains(currentLetter)) {
                uniqueLettersFound.add(currentLetter);
                return true;
            }
        }

        int[][] directions = {
                { -1, 0 }, // Arriba
                { 1, 0 }, // Abajo
                { 0, -1 }, // Izquierda
                { 0, 1 }, // Derecha
                { -1, -1 }, // Diagonal Arriba Izquierda
                { -1, 1 }, // Diagonal Arriba Derecha
                { 1, -1 }, // Diagonal Abajo Izquierda
                { 1, 1 } // Diagonal Abajo Derecha
        };

        if (0 <= direction && direction < directions.length) {
            int nextI = i + directions[direction][0];
            int nextJ = j + directions[direction][1];
            return dfs(nextI, nextJ, direction, currentChar, depth + 1);
        }

        return false;
    }

    // Método principal para ejecutar el programa
    public static void main(String[] args) {

        MutantDetector mutantDetector = new MutantDetector();
        mutantDetector.getDnaFromUser();
        char[][] dnaMatrix = mutantDetector.dnaMatrix;

        boolean isMutant = mutantDetector.isMutant(dnaMatrix);

        System.out.println("Es mutante: " + isMutant);
    }
}
